crates/re_query_cache2/src/range/results.rs

```diff
@@ -0,0 +1,747 @@
+        });
+
+        start_index..end_index
+    }
+}
+
+impl CachedRangeComponentResults {
+    /// Returns the component data as a dense vector.
+    ///
+    /// Returns an error if the component is missing or cannot be deserialized.
+    ///
+    /// Use [`PromiseResult::flatten`] to merge the results of resolving the promise and of
+    /// deserializing the data into a single one, if you don't need the extra flexibility.
+    #[inline]
+    pub fn to_dense<C: Component>(&self, resolver: &PromiseResolver) -> CachedRangeData<'_, C> {
+        // It's tracing the deserialization of an entire range query at once -- it's fine.
+        re_tracing::profile_function!();
+
+        // --- Step 1: try and upsert pending data (write lock) ---
+
+        thread_local! {
+            /// Keeps track of reentrancy counts for the current thread.
+            ///
+            /// Used to detect and prevent potential deadlocks when using the cached APIs in work-stealing
+            /// environments such as Rayon.
+            static REENTERING: RefCell<u32> = const { RefCell::new(0) };
+        }
+
+        REENTERING.with_borrow_mut(|reentering| *reentering = reentering.saturating_add(1));
+
+        if self.time_range == TimeRange::EMPTY {
+            return CachedRangeData {
+                indices: None,
+                data: None,
+                time_range: TimeRange::EMPTY,
+                front_status: PromiseResult::Ready(()),
+                back_status: PromiseResult::Ready(()),
+                reentering: &REENTERING,
+            };
+        }
+
+        let mut results = if let Some(results) = self.inner.try_write() {
+            // The lock was free to grab, nothing else to worry about.
+            Some(results)
+        } else {
+            REENTERING.with_borrow_mut(|reentering| {
+                if *reentering > 1 {
+                    // The lock is busy, and at least one of the lock holders is the current thread from a
+                    // previous stack frame.
+                    //
+                    // Return `None` so that we skip straight to the read-only part of the operation.
+                    // All the data will be there already, since the previous stack frame already
+                    // took care of upserting it.
+                    None
+                } else {
+                    // The lock is busy, but it is not held by the current thread.
+                    // Just block until it gets released.
+                    Some(self.inner.write())
+                }
+            })
+        };
+
+        if let Some(results) = &mut results {
+            // NOTE: This is just a lazy initialization of the underlying deque, because we
+            // just now finally know the expected type!
+            if results.cached_dense.is_none() {
+                results.cached_dense = Some(Box::new(FlatVecDeque::<C>::new()));
+            }
```

Comment:
this will save us from a scary `.unwrap()` later on

```suggestion
            let cached_dense = results.cached_dense.get_or_insert_with(|| {
                Box::new(FlatVecDeque::<C>::new())
            });
```

Reply:
We can't cause then we hold a ref to `results` through the entire process which makes the inner loop illegal :sob: